service/udp.go

```diff
@@ -79,130 +84,229 @@ func findAccessKeyUDP(clientIP netip.Addr, dst, src []byte, cipherList CipherLis
 	h.targetIPValidator = targetIPValidator
 }
 
-// Listen on addr for encrypted packets and basically do UDP NAT.
-// We take the ciphers as a pointer because it gets replaced on config updates.
-func (h *packetHandler) Handle(clientConn net.PacketConn) {
-	nm := newNATmap(h.natTimeout, h.m, h.logger)
-	defer nm.Close()
-	cipherBuf := make([]byte, serverUDPBufferSize)
-	textBuf := make([]byte, serverUDPBufferSize)
+func (h *associationHandler) SetTargetConnFactory(factory func() (net.PacketConn, error)) {
+	h.targetConnFactory = factory
+}
 
+type AssocationHandleFunc func(assocation net.Conn)
+
+// PacketServe listens for UDP packets on the provided [net.PacketConn], creates
+// creates and manages NAT associations, and invokes the provided `handle`
+// function for each association. It uses a NAT map to track active associations
+// and handles their lifecycle.
+func PacketServe(clientConn net.PacketConn, handle AssocationHandleFunc, metrics NATMetrics) {
+	nm := newNATmap()
+	defer nm.Close()
+	buffer := make([]byte, serverUDPBufferSize)
 	for {
-		clientProxyBytes, clientAddr, err := clientConn.ReadFrom(cipherBuf)
-		if errors.Is(err, net.ErrClosed) {
-			break
+		n, addr, err := clientConn.ReadFrom(buffer)
+		if err != nil {
+			if errors.Is(err, net.ErrClosed) {
+				break
+			}
+			slog.Warn("Failed to read from client. Continuing to listen.", "err", err)
+			continue
 		}
+		pkt := buffer[:n]
+
+		// TODO: Include server address in the NAT key as well.
+		conn := nm.Get(addr.String())
+		if conn == nil {
+			conn = &natconn{
+				Conn:   &packetConnWrapper{PacketConn: clientConn, raddr: addr},
+				readBufCh: make(chan []byte, 1),
+				bytesReadCh: make(chan int, 1),
+			}
+			metrics.AddNATEntry()
+			deleteEntry := nm.Add(addr, conn)
+			go func(conn *natconn) {
+				defer func() {
+					conn.Close()
+					deleteEntry()
+					metrics.RemoveNATEntry()
+				}()
+				handle(conn)
+			}(conn)
+		}
+		readBuf, ok := <-conn.readBufCh
+		if !ok {
+			continue
+		}
+		copy(readBuf, pkt)
+		conn.bytesReadCh <- n
+	}
+}
 
-		var proxyTargetBytes int
-		var targetConn *natconn
+// natconn adapts a [net.Conn] to provide a synchronized reading mechanism for NAT traversal.
+//
+// The application provides the buffer to `Read()` (BYOB: Bring Your Own Buffer!)
+// which minimizes buffer allocations and copying.
+type natconn struct {
+	net.Conn
 
-		connError := func() (connError *onet.ConnectionError) {
+	// readBufCh provides a buffer to copy incoming packet data into.
+	readBufCh chan []byte 
+
+	// bytesReadCh is used to signal the availability of new data and carries
+	// the length of the received packet.
+	bytesReadCh chan int
+}
+
+var _ net.Conn = (*natconn)(nil)
+
+func (c *natconn) Read(p []byte) (int, error) {
+	c.readBufCh <- p
+	n, ok := <-c.bytesReadCh
+	if !ok {
+		return 0, net.ErrClosed
+	}
+	return n, nil
+}
+
+func (c *natconn) Close() error {
+	close(c.readBufCh)
```

Comment:
This will trigger a write to a closed channel on Read, which will cause a panic.
You may need to select the `c.readBufCh` write against a read on some Done channel.


Can you add a test to reproduce that?